### PR TITLE
【AutoParallel】Add_llama_dpo_auto_parallel_intermediate

### PR DESCRIPTION
**Pull request by Xing-lil**
### PR types

New features
### PR changes

Others
### Description

Support auto_parallel in llama DPO with intermediate API.
The loss curve is same over 300 steps:
loss_comparison
The loss diff is zero:
loss_difference
The performance:
performance_comparison_filtered5

### CODE REVIEW

**Comment by jeff41404** [comment]
The description are too simple, the description needs to include a comparison of loss and performance of DPO between manual parallel and auto parallel.

**Review by jeff41404 — APPROVED** [review]
LGTM

File: llm/alignment/dpo/dpo_argument.py [hunk 1]
```diff
@@ -44,6 +44,13 @@ class DPOTrainingArguments(TrainingArguments):
         default=False,
         metadata={"help": "Whether to run benchmark by autotuner. True for from_scratch."},
     )
+    use_intermediate_api: bool = field(
+        default=False,
+        metadata={"help": "Flag indicating whether to use the intermediate API for model."},
+    )
+    max_seq_length: int = field(default=128, metadata={"help": "The maximum sequence length for model inputs."})
+    global_batch_size: int = field(default=1, metadata={"help": "Total batch size for training."})
```
Comment (DrownFish19) [inline llm/alignment/dpo/dpo_argument.py]: 这个值需要自己设置吗？

File: paddlenlp/transformers/llama/modeling_network.py [hunk 2]
```diff
@@ -1162,3 +1186,40 @@ def auto_dist_config(self, prefix=""):
         }
 
         return config
+
+
+class LlamaForCausalLMNetDPO(LlamaForCausalLMNet):
```
Comment (DrownFish19) [inline paddlenlp/transformers/llama/modeling_network.py:1191]: 这个的作用是什么？好像只创建了一个class，继承父类，但是调用方式没有任何变化。

File: paddlenlp/transformers/dpo_criterion.py [hunk 3]
```diff
@@ -233,32 +233,22 @@ def dpo_logps(
...
-                        paddle.gather(
-                            paddle.gather(per_token_logps, response_index[0], axis=0),
-                            paddle.arange(response_index[2] + offset, response_index[3], dtype=paddle.int32),
-                            axis=0,
-                        ).sum()
-                    )
-                    for response_index in response_indexs
-                ],
-                axis=0,
+            batch_idx = response_indexs[:, 0]
```
Comment (lugimzzz) [inline paddlenlp/transformers/dpo_criterion.py]: 修改这个文件是出于什么考虑？

Comment (Xing-lil) [reply paddlenlp/transformers/dpo_criterion.py]: 为了避免在动转静的时候使用while控制流，逻辑是等价的

File: paddlenlp/trl/trl_data.py [hunk 4]
```diff
@@ -183,6 +184,9 @@ def preference_collate_fn(batch, max_seq_len=None, data_type="pairwise"):
         "response_0_labels": [],
         "response_1_labels": [],
         "response_indexs": [],
+        "attention_mask": paddle.to_tensor([0], dtype="float32"),
```
Comment (DrownFish19) [inline paddlenlp/trl/trl_data.py]: 增加attention_mask后，使用flashmask返回值中同时包含attention_mask和attn_mask_startend_row_indices，是否会对DPO运行有影响？

File: paddlenlp/transformers/llama/modeling_network.py (hunk 2 above)
Comment (Xing-lil) [reply paddlenlp/transformers/llama/modeling_network.py:1191]: 动转静需要forward的参数按序排放，DPO中多了一些参数输入，故继承了一个新的forward

File: paddlenlp/trl/dpo_auto_trainer.py [hunk 5]
```diff
@@ -0,0 +1,1154 @@
...
+        if dpo_config is None:
+            raise ValueError("dpo_config is None")
+        else:
+            self.dpo_config = dpo_config
+        if not model_with_dpo_criterion:
+            if dpo_criterion is None:
+                self.dpo_criterion = AutoDPOCriterion(
+                    self.model.config, dpo_config=dpo_config, ignore_eos_token=ignore_eos_token
+                )
+                self.dpo_criterion_dy = AutoDPOCriterion(
```
Comment (DrownFish19) [inline paddlenlp/trl/dpo_auto_trainer.py:124]: 为什么这里创建两次？self.dpo_criterion_dy -> static_training -> self.args.to_static 这个调用链路是固定的吗？这样的话是不是只需要使用一个就行。

File: paddlenlp/trl/dpo_auto_trainer.py [hunk 6]
```diff
@@ -0,0 +1,1154 @@
...
+                        logits = model(**dpo_inputs)
+                        self.enable_lora(model)
+                        model.train()
+                else:
+                    with paddle.no_grad():
+                        logits = ref_model(**dpo_inputs)
+                reference_chosen_logps, reference_rejected_logps = self.dpo_criterion(logits, labels)
+            labels = labels[:3] + (reference_chosen_logps, reference_rejected_logps)
+            logits = model(**dpo_inputs)
+            policy_chosen_logps, policy_rejected_logps, sft_loss, dpo_loss, loss = self.dpo_criterion(logits, labels)
```
Comment (DrownFish19) [inline paddlenlp/trl/dpo_auto_trainer.py:231]: 这里为什么删除
```python
        # metrics
        metric_inputs = dict(
            reference_chosen_logps=reference_chosen_logps,
            reference_rejected_logps=reference_rejected_logps,
            policy_chosen_logps=policy_chosen_logps,
            policy_rejected_logps=policy_rejected_logps,
            dpo_loss=dpo_loss,
            sft_loss=sft_loss,
            train_eval=train_eval,
        )
        self.log_metric(**metric_inputs)
```

**Comment by DrownFish19** [comment]
所有文件需要pre-commit 格式化

**Comment by Xing-lil** [comment]
> 所有文件需要pre-commit 格式化

好的，感谢

File: llm/alignment/dpo/run_dpo_auto.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        data_collator=partial(
+            preference_collate_fn_auto_parallel,
+            max_seq_len=data_args.max_seq_len,
+        ),
+        ignore_eos_token=True,
+    )
+
+    if training_args.do_train:
+        train_result = trainer.train(resume_from_checkpoint=last_checkpoint)
+
```
Comment (DrownFish19) [inline llm/alignment/dpo/run_dpo_auto.py]: 好像所有的日志输出都删除了，这里自动并行没办法处理吗？
```python
        if not training_args.autotuner_benchmark and not training_args.benchmark:
            trainer.save_model(merge_tensor_parallel=training_args.tensor_parallel_degree > 1)
            trainer.log_metrics("train", train_result.metrics)
            trainer.save_metrics("train", train_result.metrics)
            trainer.save_state()
```

File: llm/alignment/dpo/run_dpo_auto.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+                "[timelog] {}: {:.2f} token/s ({}) ".format(
+                    "Effective tokens per second", effective_tokens_per_second, time.strftime("%Y-%m-%d %H:%M:%S")
+                )
+            )
+            logger.info(
+                "[timelog] {}: {:.2f} token/s ({}) ".format(
+                    "Tokens per second", total_tokens_per_second, time.strftime("%Y-%m-%d %H:%M:%S")
+                )
+            )
+
```
Comment (DrownFish19) [inline llm/alignment/dpo/run_dpo_auto.py]: 同上
```python
    if training_args.do_eval:
        eval_result = trainer.evaluate()
        trainer.log_metrics("eval", eval_result)
        trainer.save_metrics("eval", eval_result)
```

File: paddlenlp/trl/dpo_auto_trainer.py (hunk 5 above)
Comment (Xing-lil) [reply paddlenlp/trl/dpo_auto_trainer.py:124]: ref_model只计算前向所以没有做转静处理，所以用的是没有转静的dpo_criterion

File: paddlenlp/trl/dpo_auto_trainer.py (hunk 6 above)
Comment (Xing-lil) [reply paddlenlp/trl/dpo_auto_trainer.py:231]: 这里因为动转静后只返回一个loss，所以打log的时候有些限制，这块后续会增强一下

File: llm/alignment/dpo/run_dpo_auto.py (hunk 7 above)
Comment (Xing-lil) [reply llm/alignment/dpo/run_dpo_auto.py]: 同上

File: llm/alignment/dpo/run_dpo_auto.py (hunk 8 above)
Comment (Xing-lil) [reply llm/alignment/dpo/run_dpo_auto.py]: 同上

File: paddlenlp/transformers/dpo_criterion.py (hunk 3 above)
Comment (Xing-lil) [reply paddlenlp/transformers/dpo_criterion.py]: 在AutoDPOCriterion中重写了dpo_logps，自动并行用没有for的mask写法，原本动手的DPO逻辑保持不变，感谢！

File: paddlenlp/trl/trl_data.py (hunk 4 above)
Comment (Xing-lil) [reply paddlenlp/trl/trl_data.py]: 将 attention_mask 的占位初始化挪到了自动并行的collate中，验证了不影响原本动手DPO的逻辑

**Review by DrownFish19 — COMMENTED** [review]
LGTM

等待CI通过可合入

**Review by DrownFish19 — DISMISSED** [review]

**Comment by Xing-lil** [comment]
> LGTM
> 
> 等待CI通过可合入

好的，感谢

File: llm/alignment/dpo/dpo_argument.py (hunk 1 above)
Comment (Xing-lil) [reply llm/alignment/dpo/dpo_argument.py]: 已删除global_batch_size的设置

**Comment by Xing-lil** [comment]
> 所有文件需要pre-commit 格式化

已重新check pre-commit 格式化，删除冗余头文件

File: llm/alignment/dpo/dpo_argument.py [hunk 9]
```diff
@@ -44,6 +44,12 @@ class DPOTrainingArguments(TrainingArguments):
         default=False,
         metadata={"help": "Whether to run benchmark by autotuner. True for from_scratch."},
     )
+    use_intermediate_api: bool = field(
+        default=False,
+        metadata={"help": "Flag indicating whether to use the intermediate API for model."},
+    )
+    max_seq_length: int = field(default=128, metadata={"help": "The maximum sequence length for model inputs."})
```
Comment (lugimzzz) [inline llm/alignment/dpo/dpo_argument.py]: 这里max_seq_length有什么作用

File: llm/alignment/dpo/dpo_argument.py [hunk 10]
```diff
@@ -44,6 +44,12 @@ class DPOTrainingArguments(TrainingArguments):
...
+    num_hidden_layers: int = field(default=2, metadata={"help": "The number of hidden layers in the network model."})
```
Comment (lugimzzz) [inline llm/alignment/dpo/dpo_argument.py:51]: 这里num_hidden_layers需要按模型传进去？为啥不读取model config

File: llm/alignment/dpo/dpo_argument.py [hunk 11]
```diff
@@ -137,6 +143,7 @@ class DPOModelArgument:
         default=None,
         metadata={"help": "whether to fuse first up and gate proj in mlp block"},
     )
+    model_type: str = field(default="llama_dpo", metadata={"help": "Specifies the type of model to use."})
```
Comment (lugimzzz) [inline llm/alignment/dpo/dpo_argument.py]: 这个作用是什么不能通过model config.json中的architecture自动做一个映射吗？

File: paddlenlp/transformers/dpo_criterion.py [hunk 12]
```diff
@@ -304,3 +305,131 @@ def forward(
...
+                hidden_states = GatherOp.apply(hidden_states)
+                hidden_states = hidden_states.reshape(
+                    [
+                        -1,
+                        self.config.max_sequence_length,
+                        hidden_states.shape[-1],
+                    ]
+                )
+        if use_fused_head_and_loss_fn:
+            per_token_logps = -fused_head_and_loss_fn(
```
Comment (lugimzzz) [inline paddlenlp/transformers/dpo_criterion.py]: use_sparse_head_and_loss_fn、fused_head_and_loss_fn 这些都支持吗

File: llm/alignment/dpo/dpo_argument.py (hunk 9 above)
Comment (Xing-lil) [reply llm/alignment/dpo/dpo_argument.py]: 确实是冗余的，我删除一下，感谢！

File: llm/alignment/dpo/dpo_argument.py (hunk 10 above)
Comment (Xing-lil) [reply llm/alignment/dpo/dpo_argument.py:51]: 为了在自动并行训练脚本中提供灵活配置，支持按需运行小规模模型，便于在CI测试中使用。

File: llm/alignment/dpo/dpo_argument.py (hunk 11 above)
Comment (Xing-lil) [reply llm/alignment/dpo/dpo_argument.py]: 之前是为了配置不同的模型留的接口，暂时不需要，先删除了

File: paddlenlp/transformers/dpo_criterion.py (hunk 12 above)
Comment (Xing-lil) [reply paddlenlp/transformers/dpo_criterion.py]: 暂不支持，已删除相关逻辑

**Review by lugimzzz — APPROVED** [review]
LGTM
